### PR TITLE
Implement Authentication support for broker

### PR DESCRIPTION
**Pull request by warunalakshitha**
AMQP specification defined the security data layer based on Simple Authentication and Security Layer(SASL) framework. Implementation will provide Message broker to support PLAIN text SASL mechanism. Extension for pluggable authentication modules has been implemented using JAAS.

Git issue: https://github.com/wso2/message-broker/issues/61
You can define the Jaas login module in broker.yaml.

 authenticator:
  loginModule: org.wso2.broker.core.security.authentication.jaas.BrokerLoginModule

Provide extensible authentication.

## Release note
SASL based authentication support

## Automation tests
 - Integration tests
   > Integration testes added
 - Unit tests
   > Unit testes added

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: modules/broker-amqp/src/main/java/org/wso2/broker/amqp/codec/frames/ConnectionSecure.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+package org.wso2.broker.amqp.codec.frames;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import org.wso2.broker.amqp.codec.handlers.AmqpConnectionHandler;
+import org.wso2.broker.common.data.types.LongString;
+
+/**
+ * AMQP connection.secure frame.
```
Comment (a5anka) [inline modules/broker-amqp/src/main/java/org/wso2/broker/amqp/codec/frames/ConnectionSecure.java]: Shall we document the used paramaters as well similar to other frames. E.g. https://github.com/wso2/message-broker/blob/master/modules/broker-amqp/src/main/java/org/wso2/broker/amqp/codec/frames/ConnectionOpen.java

File: modules/broker-amqp/src/main/java/org/wso2/broker/amqp/codec/frames/ConnectionSecureOk.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import org.wso2.broker.amqp.codec.BlockingTask;
+import org.wso2.broker.amqp.codec.handlers.AmqpConnectionHandler;
+import org.wso2.broker.common.data.types.LongString;
+import org.wso2.broker.common.data.types.ShortString;
+
+import javax.security.sasl.SaslException;
+import javax.security.sasl.SaslServer;
+
+/**
+ * AMQP connection.secure.ok frame.
```
Comment (a5anka) [inline modules/broker-amqp/src/main/java/org/wso2/broker/amqp/codec/frames/ConnectionSecureOk.java]: Shall we document the used parameters as well similar to other frames? E.g. https://github.com/wso2/message-broker/blob/master/modules/broker-amqp/src/main/java/org/wso2/broker/amqp/codec/frames/ConnectionOpen.java

File: modules/broker-core/src/main/java/org/wso2/broker/core/Broker.java [hunk 3]
```diff
@@ -37,12 +38,24 @@
...
 
+    private final AuthenticationManager authenticationManager;
+
     public Broker(StartupContext startupContext) throws Exception {
         this.messagingEngine = new MessagingEngine(startupContext.getService(DataSource.class));
 
         BrokerServiceRunner serviceRunner = startupContext.getService(BrokerServiceRunner.class);
         serviceRunner.deploy(new BrokerAdminService());
         startupContext.registerService(Broker.class, this);
+        this.authenticationManager = new AuthenticationManager(startupContext);
```
Comment (a5anka) [inline modules/broker-core/src/main/java/org/wso2/broker/core/Broker.java]: Shall we read and pass the broker config here. It is better if we can have all the service reading logic in one place.

File: modules/broker-core/src/main/java/org/wso2/broker/core/security/authentication/AuthenticationManager.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    /**
+     * Constructor for initialize authentication manager and resgiter sasl servers for security provider
+     * mechanisms
+     */
+    public AuthenticationManager(StartupContext startupContext) throws Exception {
+        BrokerConfigProvider configProvider = startupContext.getService(BrokerConfigProvider.class);
+        BrokerConfiguration brokerConfiguration = configProvider
+                .getConfigurationObject(BrokerConfiguration.NAMESPACE, BrokerConfiguration.class);
+        BrokerConfiguration.AuthenticationConfiguration authenticator;
+        if ((authenticator = brokerConfiguration.getAuthenticator()) != null) {
```
Comment (a5anka) [inline modules/broker-core/src/main/java/org/wso2/broker/core/security/authentication/AuthenticationManager.java]: Ideally, we should have a default value in the configuration. Therefore this null check is redundant.

File: modules/broker-core/src/main/java/org/wso2/broker/core/security/authentication/AuthenticationManager.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            Configuration jaasConfig = createJaasConfig(authenticator.getLoginModule(), authenticator.getOptions());
+            Configuration.setConfiguration(jaasConfig);
+        }
+        registerSASLServers();
+    }
+
+    /**
+     * Register security provider mechanisms
+     */
+    private void registerSASLServers() {
```
Comment (a5anka) [inline modules/broker-core/src/main/java/org/wso2/broker/core/security/authentication/AuthenticationManager.java]: Shall we fix the naming convention issue?

File: modules/broker-amqp/src/main/java/org/wso2/broker/amqp/codec/frames/ConnectionStartOk.java [hunk 6]
```diff
@@ -59,8 +71,38 @@ protected void writeMethod(ByteBuf buf) {
 
     @Override
     public void handle(ChannelHandlerContext ctx, AmqpConnectionHandler connectionHandler) {
-        // TODO add authentication
-        ctx.writeAndFlush(new ConnectionTune(256, 65535, 0));
+        ctx.fireChannelRead((BlockingTask) () -> {
+            Broker broker = connectionHandler.getBroker();
+            SaslServer saslServer;
+            SaslServerBuilder saslServerBuilder = broker.getAuthenticationManager().getSaslMechanisms()
+                    .get(mechanisms.toString());
```
Comment (a5anka) [inline modules/broker-amqp/src/main/java/org/wso2/broker/amqp/codec/frames/ConnectionStartOk.java]: Shall we rename `mechanisms` to `mechanism`?

**Comment by a5anka** [comment]
Shall we document the steps required in writing a new auth module?

File: modules/broker-core/src/main/java/org/wso2/broker/core/security/authentication/AuthenticationManager.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    /**
+     * Map of SASL Server mechanisms
+     */
+    private Map<String, SaslServerBuilder> saslMechanisms = new HashMap<>();
+
+    /**
+     * Constructor for initialize authentication manager and register sasl servers for security provider mechanisms
+     */
+    public AuthenticationManager(BrokerConfiguration brokerConfiguration) throws Exception {
+        String jaasConfigPath = System.getProperty(BrokerSecurityConstants.SYSTEM_PARAM_JAAS_CONFIG);
```
Comment (a5anka) [inline modules/broker-core/src/main/java/org/wso2/broker/core/security/authentication/AuthenticationManager.java]: We should document this in some place

Comment (warunalakshitha) [reply modules/broker-core/src/main/java/org/wso2/broker/core/security/authentication/AuthenticationManager.java]: Created https://github.com/wso2/message-broker/issues/138 to comprehensive documentation on broker authentication

**Pull request merged**